src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs

```diff
@@ -775,18 +784,22 @@ private async Task AddDocumentsWithPotentialConflictsAsync(IEnumerable<Document>
                         var semanticModel = await document.GetRequiredSemanticModelAsync(_cancellationToken).ConfigureAwait(false);
                         var originalSyntaxRoot = await semanticModel.SyntaxTree.GetRootAsync(_cancellationToken).ConfigureAwait(false);
 
+                        using var _ = ArrayBuilder<RenameLocation>.GetInstance(out var locationsInDocument);
+                        foreach (var location in renameLocations)
+                        {
+                            if (location.DocumentId == documentId)
+                                locationsInDocument.Add(location);
+                        }
```

Comment:
Should we be doing a group by the document here? Otherwise if we have lots of documents and lots of rename locations, this is O(n*m)?

Reply:
Is `documentIdsToRename` already a `HashSet<Document>`?

Reply:
Oh, I misunderstood you,  you are saying renameLocations should be grouped into a Dictionary<DocumetId, ImmutableArray<RenameLocations>> before entering the loop.

Yes, that's possible.

Reply:
The existing code path seems really old 
So many places optimize

Reply:
Yeah, once you've got your stuff in I'm really tempted to stick this under a profiler and see what we discover. Seems like there's a lot of strange manipulation of data back and forth, and not just new stuff you've done.